gibberish_server/controllers/room_controller.js

```diff
@@ -24,23 +23,32 @@ class RoomController {
 
     static joinRoom(req, res) {
         // TODO: race conditionSTATE.GAME_WAITING
-        let r = roomModel.Room.deserializeRoom(roomModel.getRoomInfoAsJson(roomId));
-        if (r.state != STATE.GAME_WAITING) return "";
-        for (let p in r.players) {
-            if (p.name == nickname) {
-                return "";
-            }
-        }
-        r.players.push(new Player(nickname))
-        roomModel.saveRoom(r, client);
-        res.sen(r.players[r.length - 1].name);
+        const nickname = req.body.nickname
+        const roomId = req.body.roomId
+        roomModel.getRoomInfoAsJson(roomId)
+            .then(result => {
+                let r = roomModel.Room.deserializeRoom(result)
+                if (r.state != roomModel.STATE.GAME_WAITING) return "";
+                for (let p in r.players) {
+                    if (p.name == nickname) {
+                        return "";
+                    }
+                }
+                r.addPlayer(nickname)
+                roomModel.saveRoom(r);
+                res.send(r);
```

Comment:
note the r you send here is an object not json

Reply:
Think you have to catch the error here, in case the promise is not resolved

Reply:
@CiciliaHelena as discussed, sending as object works as well, and this skips the step of querying the database for a json version of the room
@ryanmt95 added catch blocks in the latest PRs